Add featured search page

tofumatt opened the pull request:
Fixes https://github.com/mozilla/addons-frontend/issues/1535

Adds the list of "more featured add-ons", which the mocks want.

Includes a bit of clean-up in its own commit as I didn't see the harm.

### Screenshots
![screen shot 2016-12-30 at 17 49 30](https://cloud.githubusercontent.com/assets/90871/21579502/3ec6fea2-cfab-11e6-9fb6-d7dd5d004fa0.png)
![screen shot 2016-12-30 at 17 49 14](https://cloud.githubusercontent.com/assets/90871/21579501/3ec6d74c-cfab-11e6-9f23-1e6d485c1833.png)

### Video
![dec-30-2016 17-48-55](https://cloud.githubusercontent.com/assets/90871/21579503/433f27e8-cfab-11e6-9253-980e2fea3376.gif)

kumar303 commented on src/amo/actions/featured.js:
```diff
@@ -0,0 +1,31 @@
+  };
+}
+
+export function failFeatured({ addonType }) {
```
Everything related to `FEATURED_FAILED` doesn't seem to be used in any meaningful way. I see that the action gets dispatched when the exception fails but that's it. I say we remove it all for now until we hook it up async-connect errors properly. In other words, it will probably be a YAGNI feature once we address https://github.com/mozilla/addons-frontend/issues/1521

tofumatt replied:
Fair enough. It at least shows a failure action during development but I guess that's it. I'll remove them, but in their own commit so we can always easily fetch them if wanted. 😄

kumar303 commented on src/amo/utils.js:
```diff
@@ -1,8 +1,31 @@
+
+  return featuredAPI({ api, filters })
+    .then((response) => dispatch(
+      loadFeatured({ addonType, ...response })
```
I'd rather see `entities` and `result` instead of `...response`. Since there are only two keys I don't think it saves much to use a spread.

kumar303 commented on tests/client/amo/components/TestFeaturedAddons.js:
```diff
@@ -0,0 +1,79 @@
+    return findDOMNode(findRenderedComponentWithType(renderIntoDocument(
+      <Provider store={createStore(initialState)}>
+        <I18nProvider i18n={getFakeI18nInst()}>
+          <FeaturedAddonsBase i18n={getFakeI18nInst()} {...props} />
```
do you need to pass in `i18n` again even though you use the provider?

tofumatt replied:
No, this started using `FeaturedAddonsBase` but could use `FeaturedAddons` without the `i18n` prop. We use the providers because we're rendering subcomponents and it's nice to check their data.

tofumatt replied:
Actually using `shallowRender` here might be doable and would clean this up, as removing the `i18n` prop here breaks stuff but including `FeaturedAddons` instead of `FeaturedAddonsBase` means passing stuff into the `<Provider>` from each test and is gross.

I'll try to shallowrender it if it's easy.

tofumatt replied:
I moved it to `shallowRender`, which arguably makes the properties we're checking a bit less obvious but the tests are less brittle and way fewer lines so 👍

kumar303 commented on tests/client/amo/components/TestFeaturedAddons.js:
```diff
@@ -0,0 +1,79 @@
+
+  it('throws if add-on type is not supported', () => {
+    assert.throws(() => {
+      render({ addonType: 'XUL' });
```
ha

kumar303 commented on tests/client/amo/reducers/testFeatured.js:
```diff
@@ -0,0 +1,66 @@
+import featured, { initialState } from 'amo/reducers/featured';
+
+describe('featured reducer', () => {
+  it('defaults to loading', () => {
```
it defaults to loading? That's not what it looks like from the test.

tofumatt replied:
But fortunately I only need to change the text, as it shouldn't default to loading 😄

tofumatt commented on src/amo/components/FeaturedAddons/index.js:
```diff
@@ -0,0 +1,63 @@
+}
+
+export function mapStateToProps(state) {
+  return { hasSearchParams: true, ...state.featured };
```
Totally, good call.

tofumatt commented:
All changes addressed, should be good for another r? (Sorry for the delay on getting fixes in on this one!)

kumar303 commented on src/amo/components/FeaturedAddons/index.js:
```diff
@@ -0,0 +1,66 @@
+export function mapStateToProps(state) {
+  return {
+    addonType: state.featured.addonType,
+    hasSearchParams: true,
```
could this just be set in `defaultProps` ?

tofumatt replied:
Thanks, indeed!

kumar303 commented on src/amo/components/SearchPage.js:
```diff
@@ -27,20 +22,19 @@ export default class SearchPage extends React.Component {
     const queryParams = this.props.queryParams ||
       convertFiltersToQueryParams(filters);
-    const paginator = count && hasSearchParams > 0 ? (
+    const paginator = count > 0 && hasSearchParams ? (
```
since `count` is always used like a number, maybe make the property default to 0 for sanity?

tofumatt replied:
Good call.

kumar303 commented on tests/client/amo/components/TestFeaturedAddons.js:
```diff
@@ -0,0 +1,76 @@
+
+
+describe('<FeaturedAddons />', () => {
+  const initialState = { api: { clientApp: 'android', lang: 'en-GB' } };
```
This should use [signedInApiState](https://github.com/mozilla/addons-frontend/blob/70f359d87cb70e2ecf2edf905ffedbb3e661e778/tests/client/amo/helpers.js#L45) like `{ api: signedInApiState }`

kumar303 commented on tests/client/amo/components/TestFeaturedAddons.js:
```diff
@@ -0,0 +1,76 @@
+
+  function render({ ...props }) {
+    return shallowRender(
+      <FeaturedAddonsBase i18n={getFakeI18nInst()} {...props} />
```
If you did `<FeaturedAddons store={store} ...>` (i.e. use the default component, not the base) won't that invoke `mapStateToProps` automatically?

tofumatt replied:
Because I'm using `shallowRender` (to avoid having to add `<Provider>`s and render a bunch of children), that seems to not work. I get errors like:

```
✖ renders a FeaturedAddons page with themes HTML
      Firefox 50.0.0 (Mac OS X 10.12.0)
    root.props.children is undefined
    @webpack:///tests/client/amo/components/TestFeaturedAddons.js:44:4 <- test-runner.js:126801:6
```

kumar303 commented on tests/client/amo/components/TestFeaturedAddons.js:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  it('renders a FeaturedAddons page with no add-ons set', () => {
+    const store = createStore(initialState);
```
it looks like `store =` could be moved up to a `beforeEach()`

kumar303 commented on tests/client/amo/helpers.js:
```diff
@@ -12,6 +12,7 @@ export const fakeAddon = {
     id: 123,
     license: { name: 'tofulicense', url: 'http://license.com/' },
     version: '2.0.0',
+    files: [],
```
Ha. How did this work before? I just added it too. https://github.com/mozilla/addons-frontend/pull/1586/files#diff-2e76cee7f42cc26c94ef51183f1b5f86R16

tofumatt replied:
Right?!

kumar303 replied:
You may want to merge master into your branch just to check that this change merges cleanly. However, if github isn' complaining it's probably fine.

tofumatt replied:
I rebased master onto this and pushed it here. Rebased cleanly so I think git was smart enough to realise we did the same thing ^_^

kumar303 commented on tests/client/amo/test_utils.js:
```diff
@@ -1,15 +1,50 @@
+    let ownProps;
+
+    beforeEach(() => {
+      ownProps = {
```
It doesn't look like this is shared by multiple tests anymore. I suggest moving it closer to the source for easier readability until it needs to be shared.

tofumatt replied:
👍

kumar303 commented on src/amo/utils.js:
```diff
@@ -1,10 +1,36 @@
 import { apiAddonType } from 'core/utils';
 
 
+export function fetchFeaturedAddons({ addonType, api, dispatch }) {
```
I don't understand why you need this as a separate function. Can't it just be part of `loadFeaturedAddons` ? Maybe it will become more clear in the next patch?

tofumatt replied:
I guess it can be part of it. This was just convention in previous similar utils functions (categories and search) and the smaller functions make it easier to test and such. I like the small functions personally, but if you'd rather it all together I can put them together.

Though since I'm not discretely testing it as you mentioned, I'll either do that or merge them; no use doing it halfway!

kumar303 commented on tests/client/amo/test_utils.js:
```diff
@@ -1,15 +1,50 @@
+        .withArgs({ api: {}, filters: { addonType, page_size: 25 } })
+        .returns(Promise.resolve({ entities, result }));
+
+      return loadFeaturedAddons({ store, params: ownProps.params })
```
This looks like it should be a test for `fetchFeaturedAddons` instead since it sets up mock expectations for the API call. I don't see `fetchFeaturedAddons` tested directly though and I commented above that you could just merge it into `loadFeaturedAddons` anyway and be done (unless I'm missing something).

tofumatt commented:
I think I addressed the comments; I tested `fetchFeaturedAddons` discretely though the test for `loadLandingAddons` is similar as when I tried to mock `fetchFeaturedAddons` using DI it got weird. I'll investigate more if you like though it's similar to other tests of similar methods...

If you find the methods should be combined I'll just do that. Let me know.

kumar303 commented on tests/client/amo/test_utils.js:
```diff
@@ -1,29 +1,81 @@
+    };
+  });
+
+  describe('fetchFeaturedAddons()', () => {
```
This test is exactly the same as `loadFeaturedAddons()`. I think it would make your life way easier if you merge the code for `fetchFeaturedAddons` into `loadFeaturedAddons` and only keep the test for `loadFeaturedAddons`. I know you mentioned that it follows a convention but it doesn't seem helpful in this context.

tofumatt replied:
Yeah, I see the light 😉 👍

tofumatt commented:
Combined the functions, so should be good now 😄

Thanks for being patient while I slowly worked through all the issues 😆
